framework/src/base/NonlinearSystem.C

```diff
@@ -584,48 +585,49 @@ NonlinearSystem::addNodalKernel(const std::string & kernel_name, const std::stri
 void
 NonlinearSystem::addScalarKernel(const  std::string & kernel_name, const std::string & name, InputParameters parameters)
 {
-  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); tid++)
-  {
-    MooseSharedPointer<ScalarKernel> kernel = MooseSharedNamespace::static_pointer_cast<ScalarKernel>(_factory.create(kernel_name, name, parameters, tid));
-    mooseAssert(kernel != NULL, "Not a Kernel object");
-
-    _kernels[tid].addScalarKernel(kernel);
-  }
+  MooseSharedPointer<ScalarKernel> kernel = MooseSharedNamespace::static_pointer_cast<ScalarKernel>(_factory.create(kernel_name, name, parameters));
+  mooseAssert(kernel, "Not a ScalarKernel object");
+  _scalar_kernels.addObject(kernel);
 }
 
 void
 NonlinearSystem::addBoundaryCondition(const std::string & bc_name, const std::string & name, InputParameters parameters)
 {
-  // Create a set of boundary ids
-  std::set<BoundaryID> boundary_ids;
-  const std::vector<BoundaryName> & boundary_names = parameters.get<std::vector<BoundaryName> >("boundary");
-  for (std::vector<BoundaryName>::const_iterator it = boundary_names.begin(); it != boundary_names.end(); ++it)
-    boundary_ids.insert(_mesh.getBoundaryID(*it));
 
   for (THREAD_ID tid = 0; tid < libMesh::n_threads(); tid++)
   {
+    // Create the objecte for the current thread
     parameters.set<MaterialData *>("_material_data") = _fe_problem._bnd_material_data[tid];
-
     MooseSharedPointer<BoundaryCondition> bc = MooseSharedNamespace::static_pointer_cast<BoundaryCondition>(_factory.create(bc_name, name, parameters, tid));
 
+    // Active BoundaryIDs for the object
+    const std::set<BoundaryID> & boundary_ids = bc->boundaryIDs();
+    _vars[tid].addBoundaryVar(boundary_ids, &bc->variable());
+
+    // PresetNodalBC
     MooseSharedPointer<PresetNodalBC> pnbc = MooseSharedNamespace::dynamic_pointer_cast<PresetNodalBC>(bc);
-    if (pnbc.get())
-      _bcs[tid].addPresetNodalBC(boundary_ids, pnbc);
+    if (pnbc)
+      if (tid == 0)
+        _preset_nodal_bcs.addObject(pnbc);
```

Comment:
There was threaded copies stored in the warehouse, they were not computed threaded.

https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystem.C#L822